### PR TITLE
[SPARK-17098][SQL] Fix `NullPropagation` optimizer to handle `COUNT(NULL) OVER` correctly

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

Currently, `NullPropagation` optimizer replaces `COUNT` on null literals in a bottom-up fashion. During that, `WindowExpression` is not covered properly. This PR adds the missing propagation logic.

**Before**

``` scala
scala> sql("SELECT COUNT(1 + NULL) OVER ()").show
java.lang.UnsupportedOperationException: Cannot evaluate expression: cast(0 as bigint) windowspecdefinition(ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)
```

**After**

``` scala
scala> sql("SELECT COUNT(1 + NULL) OVER ()").show
+----------------------------------------------------------------------------------------------+
|count((1 + CAST(NULL AS INT))) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)|
+----------------------------------------------------------------------------------------------+
|                                                                                             0|
+----------------------------------------------------------------------------------------------+
```
## How was this patch tested?

Pass the Jenkins test with a new test case.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Oh, I missed multi column cases. I'll fix soon.

**Comment by dongjoon-hyun** [comment]
Af first, I tried in a wrong way, so I pushed a new one which handles `NULL` propagations over WindowExpression.

**Comment by dongjoon-hyun** [comment]
It's a JVM error after python tests.
```
Could not load hsdis-amd64.so; library not loadable; PrintAssembly is disabled
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (sharedRuntime.cpp:834), pid=77834, tid=139914306811648
```

**Comment by dongjoon-hyun** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
It's `org.apache.spark.sql.execution.streaming.state.StateStoreSuite` failure.

**Comment by dongjoon-hyun** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2663,6 +2663,11 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("SPARK-17098 COUNT(NULL) on window should be 0") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: can we move this over to null-propagation.sql ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -630,6 +630,8 @@ object NullPropagation extends Rule[LogicalPlan] {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case q: LogicalPlan => q transformExpressionsUp {
+      case e @ WindowExpression(Cast(Literal(0L, _), _), _) =>
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: is there a unit test suite for NullPropagation?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Sure. I'll try.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: I can't find that. I think there does not exist.

Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Thank you for review, @rxin .

**Comment by dongjoon-hyun** [comment]
Hi, @hvanhovell .
Could you review this PR when you have some time?

**Comment by hvanhovell** [comment]
LGTM - merging this master/2.0. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for review and merging, @hvanhovell !

**Pull request closed without merging**